File: py/pe11.py
#! /usr/bin/env python3

from sys import argv

def pe11(fname, n):
    """
    What is the greatest product of four numbers
    on the same straight line in the 20 by 20 grid?
    """
    with open(fname, 'r') as f:
        grid = f.read()
    grid = grid.split('\n')
    m = 0
    size = len(grid)
    r = size - n
    for i in range(size):
        grid[i] = grid[i].split(' ')
        print(len(grid[i]))
    for x in range(size):
        for y in range(size):
            ud = lr = d0 = d1 = 1
            for i in range(n):
                lr *= int(grid[x+i][y]) if x < r else 0
                ud *= int(grid[x][y+i]) if y < r else 0
                d0 *= int(grid[x+i][y+i]) if x < r and y < r else 0
                d1 *= int(grid[x+i][y+n-i]) if x < r and y < r else 0
            if m < ud: m = ud
            if m < lr: m = lr
            if m < d0: m = d0
            if m < d1: m = d1
    print(m)

def main():
    argc = len(argv)
    if argc > 1:
        fname = argv[1]
    else:
        fname = "pe11.txt"
    if argc > 2:
        n = int(argv[2])
    else:
        n = 4

    try:
        pe11(fname, n)
    except FileNotFoundError:
        print("""file not found: %s

  python pe11.py [filepath [integer]]""" % fname)

if __name__ == "__main__":
    main()

File: cpp/src/pe11.cc
#include <pe11.h>

#include <iostream>
#include <fstream>
#include <string>

using namespace std;

uint pe11(const char* fname)
{
  int i, j, k, n;
  int nums[20][20];

  i = j = 0;
  n = -1;
  ifstream ifs(fname, ifstream::in);
  char c = ifs.get();
  while (ifs.good()) {
    if (c >= '0' && c <= '9') {
      if (n == -1) {
        n = 10 * (c - '0');
      } else {
        n += c - '0';
        nums[i][j++] = n;
        n = -1;
        if (j > 19) {
          j = 0;
          ++i;
          if (i > 19) {
            break;
          }
        }
      }
    }
    c = ifs.get();
  }
  ifs.close();

  uint ud, lr, d0, d1, m = 0;
  for (i = 0; i < 20; ++i) {
    for (j = 0; j < 20; ++j) {
      ud = lr = d0 = d1 = 1;
      for (k = 0; k < 4; ++k) {
        if (i < 16) {
          lr *= nums[i + k][j];
        } else {
          lr = 0;
        }
        if (j < 16) {
          ud *= nums[i][j + k];
        } else {
          ud = 0;
        }
        if (i < 16 && j < 16) {
          d0 *= nums[i + k][j + k];
          d1 *= nums[i + k][j + 4 - k];
        } else {
          d0 = d1 = 0;
        }
      }
      if (m < ud) {
        m = ud;
      }
      if (m < lr) {
        m = lr;
      }
      if (m < d0) {
        m = d0;
      }
      if (m < d1) {
        m = d1;
      }
    }
  }

  return m;
}

int main(int argc, char** argv)
{
  string fname = "../res/pe11.txt";

  if (argc > 1) {
    fname = argv[1];
  }

  cout << pe11(fname.c_str()) << endl;

  return 0;
}
